rollup/fees/rollup_fee.go

```diff
@@ -48,59 +45,87 @@ type StateDB interface {
 	GetBalance(addr common.Address) *big.Int
 }
 
-// CalculateL1MsgFee computes the L1 portion of the fee given
-// a Message and a StateDB
-// Reference: https://github.com/ethereum-optimism/optimism/blob/develop/l2geth/rollup/fees/rollup_fee.go
-func CalculateL1MsgFee(msg Message, state StateDB) (*big.Int, error) {
-	tx := asTransaction(msg)
+func EstimateL1DataFeeForMessage(msg Message, baseFee, chainID *big.Int, signer types.Signer, state StateDB) (*big.Int, error) {
+	if msg.IsL1MessageTx() {
+		return big.NewInt(0), nil
+	}
+
+	unsigned := asUnsignedTx(msg, baseFee, chainID)
+	// with v=1
+	tx, err := unsigned.WithSignature(signer, append(bytes.Repeat([]byte{0xff}, crypto.SignatureLength-1), 0x01))
+	if err != nil {
+		return nil, err
+	}
+
 	raw, err := rlpEncode(tx)
 	if err != nil {
 		return nil, err
 	}
 
 	l1BaseFee, overhead, scalar := readGPOStorageSlots(rcfg.L1GasPriceOracleAddress, state)
-	l1Fee := CalculateL1Fee(raw, overhead, l1BaseFee, scalar)
-	return l1Fee, nil
-}
-
-// asTransaction turns a Message into a types.Transaction
-func asTransaction(msg Message) *types.Transaction {
-	if msg.To() == nil {
-		return types.NewContractCreation(
-			msg.Nonce(),
-			msg.Value(),
-			msg.Gas(),
-			msg.GasPrice(),
-			msg.Data(),
-		)
+	l1DataFee := calculateEncodedL1DataFee(raw, overhead, l1BaseFee, scalar)
+	return l1DataFee, nil
+}
+
+// asUnsignedTx turns a Message into a types.Transaction
+func asUnsignedTx(msg Message, baseFee, chainID *big.Int) *types.Transaction {
+	if msg.AccessList() == nil {
```

Comment:
```suggestion
	if msg.AccessList() == nil && baseFee == nil {
```

Reply:
fixed in https://github.com/scroll-tech/go-ethereum/pull/351/commits/ab6c5c4d2c38375ea31cccba0e065df4c727de18